### PR TITLE
Torch symmetric - new latency optimized NVLS communication kernels for sequence parallelism 

### PR DESCRIPTION
**Pull request by sidsingh-nvidia**
# What does this PR do ?
Adding low latency NVLS multicast based all-gather and reduce-scatter kernels for sequence parallelism in inference. 
These kernels are automatically enabled if 
1. `--transformer-impl inference_optimized` is set.
2. User's pytorch supports symmetric memory
3. User is on a generation of nvlink that supports NVLS
4. the TP process group is contained within an NVLink domain.

If any of these conditions are not met, we fallback to using torch.distributed communication (NCCL).

Supports TP/SP for both mamba and gpt. 

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [x] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [x] I have added relevant unit tests
- [x] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
3. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/training/arguments.py [hunk 1]
```diff
@@ -1206,6 +1206,11 @@ def validate_args(args, defaults={}):
             args.recompute_granularity != 'full'
         ), 'recompute_granularity must not be full when CUDA Graphs are enabled.'
 
+    if args.use_inference_optimized_layers:
```
Comment (Skylion007) [inline megatron/training/arguments.py]: Does these NVShmem kernels not provide any benefit for pretraining too? Or are we just missing backward and other kernel ops?

Comment (sidsingh-nvidia) [reply megatron/training/arguments.py]: its on my bucket list to test these kernel at large batch sizes and compare them with our current best TP pretraining codepath i.e. fused AG+GEMM and GEMM+RS kernels. This might shed light on their feasibility for training.

**Comment by sidsingh-nvidia** [comment]
@santhnm2

**Comment by sidsingh-nvidia** [comment]
@wdykas for viz

**Review by shanmugamr1992 — APPROVED** [review]

**Review by duncanriach — APPROVED** [review]

**Comment by Hz188** [comment]
Hi, do you know if these two multitimem-based kernels improve training performance?
